### DISCUSSION TITLE
Setting text style for next letter in IText

### DISCUSSION THREAD

**Discussion by JesseVelden** [fabricjs/fabric.js#7146] [Q&A] [answered]
Currently I'm trying to implement a feature that allows users to change the text style for the next letter of a IText.
I found a [similar question on StackOverflow](https://stackoverflow.com/questions/38397904/fabric-js-itext-how-to-set-a-style-for-next-letter) without an helpful answer.

First I was thinking something of:
```js
object.setSelectionStyles({'fontWeight': 'bold'}, object.selectionEnd + 1)
```
But that didn't work.

And when I try to do it with: 
```js
object['fontWeight']: 'bold'
```

the entire object becomes bold (obviously). To prevent the rest of the text to also becoming bold, I could set all the individual character styles to `'fontWeight': 'normal'` where their fontWeight is not defined yet.

 But I'm not yet certain if that is a smart approach to do, and besides I don't know what method to use to loop over all the individual lines and its characters, to set their style, if it is the way to go.

**Edit:**
I got a workaround, when I add an space with object.set (which isn't rendered yet, so you don't see it), and use this:
```typescript
 if (object.isEditing && object.selectionStart === object.selectionEnd) {
      // Then we want to change the styles of the NEXT letter, but keep all the other characters the same.
      // If the hidden character ZWSP (Zero Width Space character) that we use as a flag is the last character,
      // then remove it using slice, otherwise add it. We use this to prevent in extra ZWSP characters being added when
      // this method is called twice directly after each other.
      object.set('text', object.text!.endsWith('​') ? object.text!.slice(0, -1) : (object.text += '​'));
      object.setSelectionStyles(
        { ['fontWeight']: bold ? 'normal' : 'bold' },
        object.selectionEnd!,
        object.selectionEnd! + 1
      );
}
```
it works! But honestly, it feels a bit hacky, so if anyone has better ideas, I'm open to it!
**Todo: http://fabricjs.com/docs/fabric.IText.html#insertNewStyleBlock**

**Comment by i-prikot** [comment]
Hi, @MegaCookie . Not sure that the example is suitable to solve your problem. https://codepen.io/i-prikot/pen/rNmLaYo

**Reply by JesseVelden** [reply]
Thanks for your answer, but I already have something similar implemented. I wanted to know if it is possible to have a button to change the fontWeight of the next character after the cursor, not what currently is selected.

**Comment by FabianElsmer** [comment]
I'm currently facing the same problem and looking for a way to more or less cleanly patch in styles, which only apply for the next character while editing.

Thinking of overwriting `getSelectionStyles` to handle the special case if `selectionStart` and `selectionEnd` is the same value and we have some next character styles set, then reset those next character styles once text has been entered or the cursor was changed.

Not 100% sure yet if that would work, still reading through all the `itext` code in FabricJS

Is there any simpler/cleaner solution that would accomplish the same thing? (I don't want to hack the style into an invisible next character)

**Comment by ShaMan123** [comment, marked as answer]
Why don't you flag it as `shouldChangeStylesOnInput` and listen to the itext changed event (or whatever) and change styles once input has occurred?

Sounds like a js best practice more than anything...

**Reply by FabianElsmer** [reply]
Thanks for the suggestion, this generally works :smile: 

My code looks like this to handle the case:

```
  onAdded: function () {
    this.on('text:changed', this.onTextChanged.bind(this));
  },
  onTextChanged: function () {
    if (Object.keys(this.nextCharacterStyles).length > 0) {
      // Find which character was inserted
      const previousText = this._previousText || '';
      const characterIndex = findTextInsertion(previousText, this.text);
      const insertedCharacter = this.text[characterIndex];

      // Overwrite the style for that character
      this.insertNewStyleBlock(insertedCharacter, characterIndex, [this.nextCharacterStyles]);
      this.nextCharacterStyles = {};
    }

    // Allows tracking of character changes
    this._previousText = this.text;
  },
```
Not sure if I'm using the `insertNewStyleBlock` as intended :laughing: 

A couple downsides remain, which I am totally fine with though:
- Building this on top of the `text:changed` events means I have to detect where exactly the text was changed
- The cursor will not reflect the color in `nextCharacterStyles`

Thanks for your help and time :+1:

**Reply by ShaMan123** [reply]
Keep digging
I am sure you'll manage to render the cursor with the styles

**Reply by mspanish** [reply]
Thanks @ShaMan123 - this was a great fix. My implementation:

```
// in my bold button function:

if (val === 'bold') {
    activeObject[i].set({
       shouldChangeStylesOnInput:{ fontWeight: 'bold'}
    });

}
else {
    activeObject[i].set({
       shouldChangeStylesOnInput:{ fontWeight: 'normal'}
    });
}


// in my canvas events section

canvas.on('text:changed', function (o) {
    let obj = o.target;
    if (obj.shouldChangeStylesOnInput) {
        console.log('we should be changing styles now!')
        let pos = obj.get2DCursorLocation().charIndex;
       // console.log('pos is ',pos);
      //  console.log('styles are ',obj.shouldChangeStylesOnInput);
        obj.setSelectionStyles(obj.shouldChangeStylesOnInput, pos-1, pos);
        obj.shouldChangeStylesOnInput = false;
    }
})
```
